### ISSUE TITLE
Feature request: allow panels to work like docks

### ISSUE THREAD

**Issue by CottonEaster** [linuxmint/Cinnamon#8179]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I would like Cinnamon to provide a dock-like panel. Here is what I mean: a panel that pops-up on mouse over and that has a background for its icons but which is _backgroundless_ for the part of itself that lacks icons _and which does not move icons or windows_ to accomodate it. That is, I want something that looks (and behaves) like the dock called 'plank' - and plank (on Mint) looks like this:

![image](https://user-images.githubusercontent.com/9998846/50319194-ec390280-04bd-11e9-98f6-9cf73398e131.png)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Now, at present Cinnamon _cannot_ have such a panel, without custom themes. Instead, one has the likes of this:

![image](https://user-images.githubusercontent.com/9998846/50319321-aa5c8c00-04be-11e9-824a-9756c40adcab.png)

What one sees here (immediately above) is a panel with (1) a (non-transparent) _background_ that (2) _moves_ other windows. So: CInnamon cannot give the user a panel that has a background for only the area of the screen used by the panel's icons. In short: Cinnamon cannot have a 'dock'. (There _are_ dock programmes for Linux. _All_ of them have problems.)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Users should have customisable, user-friendly version of Linux. Such I believe is the Mint ethos. Do we not have here a failure of the 'customisable, user-friendly' part?

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I apologise is the above contains any confusions. I have tried out multiple docks, multiple panel configurations, multiple panel settings, various Cinnamon extensions and multiple themes - so, as you will appreciate, it is easy to get confused.

**Comment by brownsr** [comment]
I appreciate what you are after.  We've always taken the line that if the user wants a dock, then get a dock, and we have not sought to provide a duplicate of what is out there.  The transparency bit is straightforward theming and can be done now with a suitable theme - transparent 'left' and 'right' boxes and a non-transparent 'center' box, then put the applet just in the center and leave the left and right blank.  The not moving the other windows bit is the opposite of what the panel is normally intended to do.  I seem to remember getting something similar by accident in the past, so I will have a play and post back if I can create this effect with simple theming too.  I'll post back if it works. 

Regards

Simon

**Comment by brownsr** [comment]
@CottonEaster It turned out to be simpler than I thought.  If you set a panel to autohide or intelligent hide in the panel settings then it has the behaviour of not pushing the windows out of the way,  and that sounds like what you want.  So the bare minimum is a panel with transparent sides and an opaque centre box.  I've created a theme called SimpleDock (by adapting an existing theme) that has panels set up this way on the bottom or left sides of the screen.   Top or right panels are unchanged.  It's available for installation now.  Do give me some feedback as to how well it might fit what you are asking for. You will need to set autohide/intelligent hide appropriately.  I still need to change out the screenshot.

Regards

Simon

**Comment by CottonEaster** [comment]
Simon,

Independently, I managed to implement what you suggested! The only substantial problem that I had was that window previews were overwriting part of the panel (/'dock'), for some reason. Still, I've tried your theme and in all or almost all ways it betters mine. For one thing, your version lacks the preview problem that mine had. Still, I am not wild about (1) the thick outlines on the previews or (2) the thick, semi-filled outline of the dock items. Yet, I recognise that (i) what you have done is proof-of-concept and (ii) some conspicuous way of marking the currently selected dock icon is needed. Also - though this is not directly related to the work you have done just now - I'd like to able to set 'window fade-in' time to values beyond 450ms (that's the current maximum) and up to, say, 1000ms.

Let me add that your initiative and desire to help is really rather wonderful. Thank you very much. I suspect that many users will be interested in a dock of the sort that you have created.

PS: The problem with leaving docks to third-party dock-makers is that Linux docks, at least those that can run across a range of distos/DEs, are in a bad state. Plank is the only decent such dock, I think, and it is barely maintained and it has a serious bug (it does not get out of the way of full screen windows) and it lacks proper integration into Cinnamon.

**Comment by brownsr** [comment]
Thanks.  I chose a rather individual theme of mine as a base, just because I am very familiar with it.  The changes are straightforward, so I will probably redo them in a day or so using a more 'plain vanilla' sort of theme. Which will address the points you raise.

**Comment by CottonEaster** [comment]
Brilliant. But, ah, I spoke too soon - there is a problem with the positioning of the previews:

![image](https://user-images.githubusercontent.com/9998846/50366681-1dbfd580-0573-11e9-91a0-b9e105f75300.png)

**Comment by CottonEaster** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Since I am now using your dock as my permanent theme - I have to, or else you plank, in order to have what I like - might I put in another feature request or two?

&nbsp;&nbsp;&nbsp;&nbsp;First, consider this:

![image](https://user-images.githubusercontent.com/9998846/50375258-88fcbc80-05f2-11e9-876e-437908350fbb.png)

The partial box around the settings and terminal icon seems visually jarring to me. Could it be either removed (leaving the background colour to indicate a running app) or made into a full box (a square rather than a line?).

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Second, and less importantly: the scroll par on the Mint Menu (not on my system generally) seems a bit wide:

![image](https://user-images.githubusercontent.com/9998846/50375291-ff012380-05f2-11e9-867e-2d6f895cd956.png)

Any chance of reducing the width by, say, 10%? (Still, I have overlay scrollbars disabled. I could enable them.)

**Comment by brownsr** [comment]
There is actually a full box, but it's not showing that clearly because of the closeness of the panel.  I will see if there is anything I can do there.  Ditto the slider.

**Comment by CottonEaster** [comment]
Thank you.

*Please note that your theme introduces a subtle horizontal line across* **part** - *the right half - of the top of my screen*. Strange! The effect is just about visible, if you zoom in, [in the image I have posted here](https://app.box.com/s/dtiwj1dmv4ihewf9qqtuheumnh132dhd). The white border atop and below the green is an artefact of my attempts (in GIMP) to make the problem (the line) show up; look instead within the green block. I coloured my black background green (in GIMP) and you should be able to see that towards the right side of the image there is black horizontal gap in the green.

The effect may be easier to see, though, if you simply switch from your theme to some other desktop themes and back again.

**Comment by CottonEaster** [comment]
I've also had the panel, set to 'intelligently hide', totally stop hiding. The effect persisted when I changed the setting to 'autohide'. Restarting Cinnamon fixed the problem. This problem may or may not be related to the dock theme.

**Comment by brownsr** [comment]
I can't see that panel settings are associated with a theme.  If there is something persisting when it should not be the cause probably lies elsewhere.  If you can pin it down then you could log an issue, there has been quite a lot of change around the panel and window handling in this cycle.

As far as the line is concerned I can't reproduce it.  A useful trick if you don't already know it is to turn on the zoom in the accessibility controls, and then use alt-mousewheel to zoom in and this means that you can see in forensic detail.  There are two possible causes.  The first is theme related, there are a lot of border hover effects and this might be one of those activating over an invisible applet, if so I would expect to see a more obvious effect.  I tried with the separator and spacer applets to see if it was those, but I could not reproduce.  I will scan the CSS to see if there are any stray hover elements that are not intended though. Check where your mouse is if you see this again, if it is nearby then move it away and see if the effect persists.  The second is that it might be graphical artefacting because of optimisations in the window handler this development cycle.  We did actually get a couple of these earlier on in development, the symptom was a tiny black (presumably unpainted) spot towards the edge, but I seem to remember that was for windows rather than panels.

**Comment by CottonEaster** [comment]
Thank you for your work.

Didn't know about zoom. I'ved used it to get a partial screenshot in which the line is just about visible - if you zoom in.

![screenshot from 2018-12-23 11-45-32](https://user-images.githubusercontent.com/9998846/50383302-b441e300-06a8-11e9-930c-451482166f0c.png)

The position of the mouse does not affect the line.
Some relevant settings on my system:

![image](https://user-images.githubusercontent.com/9998846/50383313-f9feab80-06a8-11e9-9276-eeea2b86daee.png)

**Comment by brownsr** [comment]
I'm afraid I can't see anything in your picture, and still can't reproduce locally, so I won't pursue that further. Sometimes my own screen gives subtle bands, either down to the screen itself or to the GPU.

**Comment by CottonEaster** [comment]
Let me try again - [here is a video](https://app.box.com/s/dtiwj1dmv4ihewf9qqtuheumnh132dhd), using zoom. One might call the line I see a subtle band, but I did not have it before 19.1.

**Comment by CottonEaster** [comment]
I have found a fix for the problem of the line at the top-right of the screen and made other modifications to my liking to the theme. I'll post here about those modifications after Christmas.

**Comment by CottonEaster** [comment]
Here ([gist](https://gist.github.com/CottonEaster/ec2b5c44c545233f85f4bc07af056747)) is my modified version of the css file. As said, that version fixes the line problem and makes some other changes. I have commented the changes, within the file.

Here are some remaining problems.

1. Sometimes the background remains - on the panel, on the right - for an icon even when the icon (from a running application) has disappeared (because the application ceased running).
3. At all times there is a small amount of extra background to the panel, on the right.
5. The green fill for the currently selected (open and foregrounded) application/icon does not seem to extend to the full rightmost width of the icon. I can send a picture.

I cleaned up the spacing a bit too, and removed a double semicolon.

Still, I do like this dock.

**Comment by brownsr** [comment]
Thanks, I've adopted your changes.  I've also changed the hover gradient rule to avoid that dark grey zone.    I suspect there may be some hard-coded padding in the grouped window list, as if I pull in another applet to the right then I don't see that bit of extra background.

I'm closing this issue now. Hope that is OK. I think you can reopen if you disagree.

Regards

Simon

**Comment by LinuxOnTheDesktop** [comment]
Cf. #9100 (and 5146).
